### PR TITLE
[#12473]  Add steps on how to use the masquerade mode in developer guide

### LINKED ISSUE #12473: Add steps on how to enter masquerade mode in developer guide
As an extension to #12466, we can add steps in the [developer guide](https://teammates.github.io/teammates/development.html#logging-in-to-a-teammates-instance) to make it clearer on how we can enter masquerade mode. This discussion #12433 outlines the steps and would be helpful.

This can be a new subsection under `Logging in to a TEAMMATES instance`, after `Logging in without UI`
image

### PR DESCRIPTION
**Pull request by theKelvincode**
Fixes #12473 

**Outline of Solution**
1. Defined what a masquerade mode is.
2. Itemised steps to perform before using the masquerade mode.
3. Listed the steps on how to use the masquerade mode.

### CODE REVIEW

File: docs/development.md [hunk 1]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
 GET http://localhost:8080/logout
 ```
 
+### How to enter masquerade mode
+Masquerade mode is a feature that enables the admin create mock instructors and students for testing and other purposes.
```
Comment (weiquu) [inline docs/development.md:190]:
```suggestion
Masquerade mode is a feature that enables the admin to create and log in as mock instructors and students for testing and other purposes.
```

File: docs/development.md [hunk 2]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+- To be able to use the masquerade mode, do the following :
```
Comment (weiquu) [inline docs/development.md:191]:
```suggestion
- To prepare for entering masquerade mode, do the following:
```

File: docs/development.md [hunk 3]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+    - Log in as an administrator to `http://localhost:8080/web/admin/home`.
+    - Create new instructor with by filling the Name, Email, Institution fields in the form
+      provided and submit.
```
Comment (weiquu) [inline docs/development.md:194]:
```suggestion
    - Create a new instructor by filling the Name, Email, and Institution fields in the form
      provided and submit.
```

File: docs/development.md [hunk 4]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+    - After adding instructor, copy the join link and paste in another browser.
```
Comment (weiquu) [inline docs/development.md:195]:
```suggestion
    - After adding the instructor, copy the join link and paste it in another browser.
```

File: docs/development.md [hunk 5]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+    - In the new browser, log in with the new instructor's email
+      (not the default admin email) and register to course.
```
Comment (weiquu) [inline docs/development.md:197]:
```suggestion
    - In the new browser, log in with the new instructor's email
      (not the default admin email) and click to register for the course.
```

File: docs/development.md [hunk 6]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+    - Now there is a new instructor with its own google id that can be used to login.
```
Comment (weiquu) [inline docs/development.md:198]: Suggest removing this entirely, or changing it to "Now, we have a new account that we can masquerade as." In any case, do refrain from using "google id" as mentioned in [this comment](https://github.com/TEAMMATES/teammates/issues/12466#issuecomment-1578848527)

File: docs/development.md [hunk 7]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+- To masquerade as an instructor :
```
Comment (weiquu) [inline docs/development.md:199]:
```suggestion
- To masquerade as an instructor:
```

File: docs/development.md [hunk 8]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+    - Use the instructor's URL at /web/instructor/home with the instructor's email id as a parameter eg. `http://localhost:8080/web/instructor/home?user=kelvin@gmail.com`
```
Comment (weiquu) [inline docs/development.md:200]:
```suggestion
    - Use the instructor's URL at /web/instructor/home with the instructor's email as a parameter, e.g. `http://localhost:8080/web/instructor/home?user=kelvin@gmail.com`
```

File: docs/development.md [hunk 9]
```diff
@@ -186,6 +186,21 @@ To "log out", submit the following API call:
...
+- To masquerade as a student:
+    - Use the student's URL at /web/student/home appended with the student's email id eg. `http://localhost:8080/web/student/home?user=janethestudent@gmail.com`
```
Comment (weiquu) [inline docs/development.md:202]:
```suggestion
    - Use the student's URL at /web/student/home appended with the student's email, e.g. `http://localhost:8080/web/student/home?user=janethestudent@gmail.com`
```

**Review by weiquu — CHANGES_REQUESTED** [review]
Thanks @theKelvincode, mainly formatting issues and some small wording changes for clarity

File: docs/development.md (hunk 6 above)
Comment (theKelvincode) [reply docs/development.md:198]: I think your suggestion of rewording it is valid considering that emails for accounts are no longer restricted to google emails.

**Comment by weiquu** [comment]
Hi @theKelvincode, thanks for the changes! Just one last comment:

For masquerading as an instructor the instructions are: "[...] with the instructor's email as a parameter", while for a student the instructions are: "[...] appended with the student's email". Since both methods are using `?user=`, can we standardise the language used to minimise confusion? You can either change one to the other, or combine them together, e.g. "[...] with the instructor's email appended as a parameter".

**Review by weiquu — APPROVED** [review]
LGTM!

File: docs/development.md [hunk 10]
```diff
@@ -186,6 +186,18 @@ To "log out", submit the following API call:
...
+Masquerade mode is a feature that enables the admin to create and log in as mock instructors and students for testing and other purposes.
+- To prepare for entering masquerade mode, do the following:
+    - Log in as an administrator to `http://localhost:8080/web/admin/home`.
+    - Create a new instructor by filling the Name, Email, and Institution fields in the form
+      provided and submit.
+    - Now, we have a new account that we can masquerade as.
+- To masquerade as an instructor:
+    - Use the instructor's URL at /web/instructor/home with the instructor's email appended as a parameter, e.g. `http://localhost:8080/web/instructor/home?user=kelvin@gmail.com`
+- To masquerade as a student:
+    - Use the student's URL at /web/student/home with the student's email appended as a parameter, e.g. `http://localhost:8080/web/student/home?user=janethestudent@gmail.com`
```
Comment (domlimm) [inline docs/development.md:199]:
```suggestion
    - Use the student's URL at /web/student/home with the student's email appended as a query parameter, e.g. `http://localhost:8080/web/student/home?user=janethestudent@gmail.com`.
```

File: docs/development.md [hunk 11]
```diff
@@ -186,6 +186,18 @@ To "log out", submit the following API call:
...
 
+### How to enter masquerade mode
+Masquerade mode is a feature that enables the admin to create and log in as mock instructors and students for testing and other purposes.
+- To prepare for entering masquerade mode, do the following:
+    - Log in as an administrator to `http://localhost:8080/web/admin/home`.
+    - Create a new instructor by filling the Name, Email, and Institution fields in the form
+      provided and submit.
+    - Now, we have a new account that we can masquerade as.
+- To masquerade as an instructor:
+    - Use the instructor's URL at /web/instructor/home with the instructor's email appended as a parameter, e.g. `http://localhost:8080/web/instructor/home?user=kelvin@gmail.com`
```
Comment (domlimm) [inline docs/development.md:197]:
```suggestion
    - Use the instructor's URL at /web/instructor/home with the instructor's email appended as a query parameter, e.g. `http://localhost:8080/web/instructor/home?user=kelvin@gmail.com`.
```

File: docs/development.md [hunk 12]
```diff
@@ -186,6 +186,18 @@ To "log out", submit the following API call:
...
 ```
 
+### How to enter masquerade mode
+Masquerade mode is a feature that enables the admin to create and log in as mock instructors and students for testing and other purposes.
+- To prepare for entering masquerade mode, do the following:
+    - Log in as an administrator to `http://localhost:8080/web/admin/home`.
+    - Create a new instructor by filling the Name, Email, and Institution fields in the form
+      provided and submit.
+    - Now, we have a new account that we can masquerade as.
+- To masquerade as an instructor:
```
Comment (domlimm) [inline docs/development.md:196]:
```suggestion
- To masquerade as an _instructor_:
```

File: docs/development.md [hunk 13]
```diff
@@ -186,6 +186,18 @@ To "log out", submit the following API call:
...
+    - Use the instructor's URL at /web/instructor/home with the instructor's email appended as a parameter, e.g. `http://localhost:8080/web/instructor/home?user=kelvin@gmail.com`
+- To masquerade as a student:
```
Comment (domlimm) [inline docs/development.md:198]:
```suggestion
- To masquerade as a _student_:
```

**Review by domlimm — CHANGES_REQUESTED** [review]
Sorry for getting back so late.

Couple of nits.

File: docs/development.md [hunk 14]
```diff
@@ -186,6 +186,18 @@ To "log out", submit the following API call:
 GET http://localhost:8080/logout
 ```
 
+### How to enter masquerade mode
+Masquerade mode is a feature that enables the admin to create and log in as mock instructors and students for testing and other purposes.
```
Comment (domlimm) [inline docs/development.md:190]:
```suggestion
Masquerade mode is a feature that enables the admin to create and log in as mock instructors and students for testing and situations where masquerading as another user is applicable.
```

**Comment by domlimm** [comment]
@theKelvincode I'm discussing with @weiquu on the masquerade mode. Might need you to make some changes after. Thank you for your patience and contribution! 😃

File: docs/development.md [hunk 15]
```diff
@@ -186,6 +186,18 @@ To "log out", submit the following API call:
...
+Masquerade mode is a feature that enables the admin to create and log in as mock instructors and students for testing and situations where masquerading as another user is applicable.
+- To prepare for entering masquerade mode, do the following:
+    - Log in as an administrator to `http://localhost:8080/web/admin/home`.
+    - Create a new instructor by filling the Name, Email, and Institution fields in the form
+      provided and submit.
+    - Now, we have a new account that we can masquerade as.
+- To masquerade as an _instructor_:
+    - Use the instructor's URL at /web/instructor/home with the instructor's email appended as a query parameter, e.g. `http://localhost:8080/web/instructor/home?user=kelvin@gmail.com`.
+- To masquerade as a _student_:
+    - Use the student's URL at /web/student/home with the student's email appended as a query parameter, e.g. `http://localhost:8080/web/student/home?user=janethestudent@gmail.com`.
```
Comment (domlimm) [inline docs/development.md:199]:
```suggestion
### Masquerading as another user

Masquerade mode is a feature that enables the admin to log in as mock instructors/students for testing and situations where masquerading as another user is applicable.

Essentially, when you are logged into the administrator account, you will be able to conveniently and directly access the other instructor/student accounts. When in masquerade mode, you should see a `(M)` at the top right of the page you are on.

**Note**:
If you decide to use port `8080` for the following steps below, make sure you have run `npm run build` in your project root. For more information, refer to [Building front-end files](#building-front-end-files). Else, you may use port `4200` which is TEAMMATES' frontend URL.

To prepare for _masquerade mode_, do the following:

- Log in as an administrator via http://localhost:4200/web/admin/home.
  - For more information, refer to the [Logging in to a TEAMMATES instance](#logging-in-to-a-teammates-instance) section above in this document.
- Create a new instructor by filling in the Name, Email, and Institution fields in the form provided. Once done, click on _Add Instructor_, followed by _Add_ under the **Action** column.
  - Copy the link address from the **_join link_**.
- Log out from the administrator account. Use the link you copied in the previous step and paste it into your browser's search bar. Enter the new account's email and register for the course.
  - Now, the registration is !!completed!! and we have a new account that we can masquerade as.
- Lastly, log out from the current account and log back in as an administrator and follow the rest of the steps below.

To masquerade as an **_instructor_**:

- Use the instructor's path `/web/instructor/home` with the instructor's email appended as a query parameter, e.g., `http://localhost:4200/web/instructor/home?user=kelvin@gmail.com`.
- For convenience, http://localhost:4200/web/instructor/home?user=INSTRUCTOR_EMAIL. Double click on `INSTRUCTOR_EMAIL` to replace it with the instructor's email.

To masquerade as a **_student_**:

- Use the student's path `/web/student/home` with the student's email appended as a query parameter, e.g., `http://localhost:4200/web/student/home?user=janethestudent@gmail.com`.
- For convenience, http://localhost:4200/web/student/home?user=STUDENT_EMAIL. Double click on `STUDENT_EMAIL` to replace it with the student's email.
```

**Review by domlimm — CHANGES_REQUESTED** [review]
@theKelvincode Sorry for the delay. Let's go with the suggested change.

**Comment by domlimm** [comment]
@theKelvincode Thank you for making the change and your patience! Let me do a final sanity check locally before we merge this in. Will get it done by today, 17 Jul. Thanks Kelvin!

**Comment by theKelvincode** [comment]
> @theKelvincode Thank you for making the change and your patience! Let me do a final sanity check locally before we merge this in. Will get it done by today, 17 Jul. Thanks Kelvin!

You're welcome. I enjoyed working on this and I look forward to making more contributions.

**Review by domlimm — APPROVED** [review]
LGTM! Thank you for your contribution, and we look forward to reviewing your PRs!

**Pull request merged**